zone-setup/src/bin/zone-setup.rs

```diff
@@ -4,285 +4,185 @@
-use std::io::Write;
+use std::io::Write as _;
 use std::net::{IpAddr, Ipv4Addr, Ipv6Addr};
 use std::os::unix::fs::chown;
-use std::path::Path;
+use std::path::PathBuf;
 use uzers::{get_group_by_name, get_user_by_name};
 
-pub const HOSTS_FILE: &str = "/etc/inet/hosts";
-pub const CHRONY_CONFIG_FILE: &str = "/etc/inet/chrony.conf";
-pub const LOGADM_CONFIG_FILE: &str = "/etc/logadm.d/chrony.logadm.conf";
-pub const ROOT: &str = "root";
-pub const SYS: &str = "sys";
-
-pub const COMMON_NW_CMD: &str = "common-networking";
-pub const OPTE_INTERFACE_CMD: &str = "opte-interface";
-pub const CHRONY_SETUP_CMD: &str = "chrony-setup";
-
-fn parse_ip(s: &str) -> anyhow::Result<IpAddr> {
-    if s == "unknown" {
-        return Err(anyhow!("ERROR: Missing input value"));
-    };
-    s.parse().map_err(|_| anyhow!("ERROR: Invalid IP address"))
+#[derive(Debug, Parser)]
+struct ZoneSetup {
+    #[command(subcommand)]
+    command: ZoneSetupCommand,
 }
 
-fn parse_ipv4(s: &str) -> anyhow::Result<Ipv4Addr> {
-    if s == "unknown" {
-        return Err(anyhow!("ERROR: Missing input value"));
-    };
-    s.parse().map_err(|_| anyhow!("ERROR: Invalid IPv4 address"))
+#[derive(Debug, Subcommand)]
+enum ZoneSetupCommand {
+    /// Sets up common networking configuration across all zones
+    CommonNetworking(CommonNetworkingArgs),
+    /// Sets up OPTE interface
+    OpteInterface(OpteInterfaceArgs),
+    /// Sets up Chrony configuration for NTP zone
+    ChronySetup(ChronySetupArgs),
 }
 
-fn parse_ipv6(s: &str) -> anyhow::Result<Ipv6Addr> {
-    if s == "unknown" {
-        return Err(anyhow!("ERROR: Missing input value"));
-    };
-    s.parse().map_err(|_| anyhow!("ERROR: Invalid IPv6 address"))
+// Some of our `String` cli arguments may be set to the literal "unknown" if
+// sled-agent hasn't populated our SMF config correctly. Reject this at argument
+// parsing time.
+fn parse_string_rejecting_unknown(s: &str) -> anyhow::Result<String> {
+    if s != "unknown" {
+        Ok(s.to_string())
+    } else {
+        Err(anyhow!("missing input value"))
+    }
 }
 
-fn parse_datalink(s: &str) -> anyhow::Result<String> {
-    if s == "unknown" {
-        return Err(anyhow!("ERROR: Missing data link"));
-    };
-    s.parse().map_err(|_| anyhow!("ERROR: Invalid data link"))
+#[derive(Debug, Args)]
+struct CommonNetworkingArgs {
+    #[arg(short, long, value_parser = parse_string_rejecting_unknown)]
+    datalink: String,
+    #[arg(short, long)]
+    gateway: Ipv6Addr,
+    #[arg(short, long)]
+    static_addr: Ipv6Addr,
 }
 
-fn parse_opte_iface(s: &str) -> anyhow::Result<String> {
-    if s == "unknown" {
-        return Err(anyhow!("ERROR: Missing OPTE interface"));
-    };
-    s.parse().map_err(|_| anyhow!("ERROR: Invalid OPTE interface"))
+#[derive(Debug, Args)]
+struct OpteInterfaceArgs {
+    #[arg(short, long, value_parser = parse_string_rejecting_unknown)]
+    interface: String,
+    #[arg(short, long)]
+    gateway: Ipv4Addr,
```

Comment:
Hmm. I thought the `--opte-` prefixes on all of these arguments was superfluous since they're only valid to the `opte-interface` subcommand. I'm 100% on board with adding descriptions, if you think that's enough; it's also easy to put the prefixes back if you think that would be better.

Reply:
There weren't any descriptions on these flags before - what should they be? :sweat_smile:

Reply:
> Hmm. I thought the --opte- prefixes on all of these arguments was superfluous since they're only valid to the opte-interface subcommand

I totally agree with this. I guess this all comes from my own experience working with the zones where several things were called "gateway" but they all meant different things and it was a bit confusing to know which went where. The quickest solution I found was to differentiate this one by just calling it "opte-gateway", but I totally get that it looks weird as a flag.

I'm not specially attached to any solution, but perhaps keeping it as `gateway` and writing a long description with something simple like `gateway is unique to OPTE as it is for external connectivity using boundary services` makes sense? Happy to go with whatever you think is best